### PR TITLE
smart deploys

### PR DESCRIPTION
**Pull request by jkminneti**
## what's this
- wait for rollouts to finish
- rollback on failures
- delete removed resources
- store previously applied state for manual rollbacks

## testing
~`tuber deploy potatoes-broken` - current build will create a new deployment and service. the deployment will fail while rolling out after the apply, and both will be deleted. - branch is `master-broken` if you want to throw other stuff at it~

lemme know if you want to try it, testing some stuff rn

## todo
- argo support

## open questions
- what's the timeout for rollout rollbacks? does that need to be configurable as well?
- should roles and rolebindings be tuberized or not?

## pseudocode
1. identify incoming .tuber yamls by name and kind
2. pull last tuber-applied resources from `configmap tuber-state-(appname) -n appname`
3. apply non-workloads
  -- on failure, reapply matching configs from state, delete newly applied configs
4. apply workloads
  -- on apply failure, roll back applied configs and workloads, delete newly applied
  -- watch rollout-y workloads for success
  -- on watch failure,  roll back applied configs and workloads, delete newly applied
5. update state configmap
  -- save last tuber-applied stuff under previousState so we can implement a rollback command
  -- save just-applied stuff for next deploy
  -- if something was in the last state but not in what was just applied, delete it

### CODE REVIEW

**Review by Hbbb — COMMENTED** [review]
I need to take another pass over this when it's ready for review. Initially though I like the idea. I'd like to talk about the purpose of this a bit more though.

File: pkg/core/releaser.go [hunk 1]
```diff
@@ -0,0 +1,450 @@
...
+	if resource.supportsRollback() && !resource.isRollout() {
+		err := k8s.RolloutStatus(resource.kind, resource.name, r.app.Name)
+		if err != nil {
+			errors <- rolloutError{err: err, resource: resource}
+		}
+	}
+	wg.Done()
+}
+
+func (r releaser) rollback(appliedResources []appResource, cachedResources []appResource) ([]appResource, []error) {
```
[the remark this answers is not in the document]
Comment (jaysonesmith) [reply pkg/core/releaser.go:334]: Can do this on 382, too.

File: pkg/core/releaser.go [hunk 2]
```diff
@@ -0,0 +1,450 @@
...
+	if applied.supportsRollback() {
+		if applied.isRollout() {
+			// TODO: add actual argo support
+			err = k8s.Apply(cached.contents, r.app.Name)
+		} else {
+			err = k8s.RolloutUndo(applied.kind, applied.name, r.app.Name)
+		}
+	} else {
+		err = k8s.Apply(cached.contents, r.app.Name)
+	}
```
Comment (jaysonesmith) [inline pkg/core/releaser.go:374]: Is this written this way for your TODO around argo support? If not we could simplify?

```suggestion
	if applied.supportsRollback() && !applied.isRollout() {
			err = k8s.RolloutUndo(applied.kind, applied.name, r.app.Name)
	} else {
		err = k8s.Apply(cached.contents, r.app.Name)
	}
```

File: pkg/core/interpolate.go [hunk 3]
```diff
@@ -31,3 +31,18 @@ func interpolate(templateString string, data map[string]string) (interpolated []
...
+	DefaultGateway string
+	DefaultHost    string
+}
+
+func releaseData(digest string, app *TuberApp, clusterData *ClusterData) (data map[string]string) {
+	return map[string]string{
+		"tuberImage":            digest,
+		"clusterDefaultGateway": clusterData.DefaultGateway,
+		"clusterDefaultHost":    clusterData.DefaultHost,
+		"tuberAppName":          app.Name,
```
Comment (jaysonesmith) [inline pkg/core/interpolate.go:46]: I know it's unlikely to happen, but if these were sent in as nil, wouldn't we get NPE's?

**Review by jaysonesmith — APPROVED** [review]
Nothing immediately looks broken! 🎉

File: pkg/core/interpolate.go (hunk 3 above)
Comment (jkminneti) [reply pkg/core/interpolate.go:46]: that means this function would have to return an error, which I don't really want to do. 

I wonder about the convention here, and where the nil checks end. Look at functions that take a `*sync.WaitGroup` - i've never seen someone nil check those

I think if anything this points out that we should have a `recovery()` deferred all the way back up in the pubsub listener

File: pkg/core/releaser.go [hunk 4]
```diff
@@ -0,0 +1,450 @@
...
+	return genericized, stateResource, nil
+}
+
+type metadata struct {
+	Name   string
+	Labels map[string]string
+}
+
+type parsedResource struct {
+	ApiVersion string
```
[the remark this answers is not in the document]
Comment (jkminneti) [reply pkg/core/releaser.go:215]: goland's unconventional linter strikes again :/ fixing

File: pkg/core/releaser.go [hunk 5]
```diff
@@ -0,0 +1,450 @@
...
+			configs = append(configs, resource)
+		}
+	}
+	return workloads, configs, nil
+}
+
+func (r releaser) apply(resources []appResource) ([]appResource, error) {
+	var applied []appResource
+	for _, resource := range resources {
+		var err error
```
[the remark this answers is not in the document]
Comment (jkminneti) [reply pkg/core/releaser.go:259]: nope! looks like an artifact from an inner loop that i later removed

File: pkg/core/releaser.go [hunk 6]
```diff
@@ -0,0 +1,450 @@
...
+	case err := <-errors:
+		scope, logger := err.resource.scopes(r)
+		return ErrorContext{err: err.err, scope: scope, logger: logger, context: "watch workload"}
+	}
+}
+
+func (r releaser) watchRollback(appliedWorkloads []appResource) []error {
+	var wg sync.WaitGroup
+	errorChan := make(chan rolloutError)
+	done := make(chan bool)
```
[the remark this answers is not in the document]
Comment (jkminneti) [reply pkg/core/releaser.go:296]: yeaaaah i know. but this time it makes sense. and i _think_ i set up the receivers and waitgroups properly - though.. unconventionally.

File: pkg/core/releaser.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (jkminneti) [reply pkg/core/releaser.go:334]: i've had this conversation with harrison too - i'm not a fan of that feature. just feels less explicit and less easy to refactor, all to just save line length

File: pkg/core/releaser.go (hunk 2 above)
Comment (jkminneti) [reply pkg/core/releaser.go:374]: totally for the TODO - which will be resolved before this exits draft status if all goes according to plan

File: pkg/events/processor.go [hunk 7]
```diff
@@ -113,13 +112,12 @@ func (p Processor) deploy(event event, app *core.TuberApp) {
 		deployLogger.Info("prerelease complete")
 	}
 
-	releaseIDs, err := core.ReleaseTubers(releaseYamls, app, event.digest, p.clusterData)
+	startTime := time.Now()
+	err = core.Release(deployLogger, errorScope, releaseYamls, app, event.digest, p.clusterData)
 	if err != nil {
-		deployLogger.Warn("failed release", zap.Error(err))
-		report.Error(err, errorScope.WithContext("release"))
+		deployLogger.Info("release failed")
```
[the remark this answers is not in the document]
Comment (jkminneti) [reply pkg/events/processor.go:118]: ugh it's inconsistent all over this file isnt it. fixing

**Pull request merged**